// decorator-inspectors/dist/AccessorInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessorSpy = exports.AccessorInspector = exports.LogAccessorInspector = void 0;
require("reflect-metadata");
function LogAccessorInspector(target, propertyKey, descriptor) {
    console.log(AccessorInspector(target, propertyKey, descriptor));
}
exports.LogAccessorInspector = LogAccessorInspector;
function AccessorInspector(target, propertyKey, descriptor) {
    return {
        target, propertyKey, descriptor,
        ownKeys: Reflect.ownKeys(target),
        design: {
            type: Reflect.getMetadata("design:type", target, propertyKey),
            paramtypes: Reflect.getMetadata("design:paramtypes", target, propertyKey),
            returntype: Reflect.getMetadata("design:returntype", target, propertyKey)
        }
    };
}
exports.AccessorInspector = AccessorInspector;
function AccessorSpy() {
    return function (target, propertyKey, descriptor) {
        const originals = {
            get: descriptor.get,
            set: descriptor.set
        };
        if (originals.get) {
            descriptor.get = function () {
                const ret = originals.get.call(this);
                console.log(`AccessorSpy get ${String(propertyKey)}`, ret);
                return ret;
            };
        }
        if (originals.set) {
            descriptor.set = function (newval) {
                console.log(`AccessorSpy set ${String(propertyKey)}`, newval);
                originals.set.call(this, newval);
            };
        }
    };
}
exports.AccessorSpy = AccessorSpy;
