// decorator-inspectors/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAccessorDecorator = exports.isMethodDecorator = exports.isParameterDecorator = exports.isPropertyDecorator = exports.isClassDecorator = exports.LogParameterInspector = exports.ParameterInspector = exports.LogMethodInspector = exports.MethodInspector = exports.LogPropertyInspector = exports.PropertyInspector = exports.AccessorSpy = exports.LogAccessorInspector = exports.AccessorInspector = exports.LogClassInspector = exports.ClassInspector = void 0;
var ClassInspector_js_1 = require("./ClassInspector.js");
Object.defineProperty(exports, "ClassInspector", { enumerable: true, get: function () { return ClassInspector_js_1.ClassInspector; } });
Object.defineProperty(exports, "LogClassInspector", { enumerable: true, get: function () { return ClassInspector_js_1.LogClassInspector; } });
var AccessorInspector_js_1 = require("./AccessorInspector.js");
Object.defineProperty(exports, "AccessorInspector", { enumerable: true, get: function () { return AccessorInspector_js_1.AccessorInspector; } });
Object.defineProperty(exports, "LogAccessorInspector", { enumerable: true, get: function () { return AccessorInspector_js_1.LogAccessorInspector; } });
Object.defineProperty(exports, "AccessorSpy", { enumerable: true, get: function () { return AccessorInspector_js_1.AccessorSpy; } });
var PropertyInspector_js_1 = require("./PropertyInspector.js");
Object.defineProperty(exports, "PropertyInspector", { enumerable: true, get: function () { return PropertyInspector_js_1.PropertyInspector; } });
Object.defineProperty(exports, "LogPropertyInspector", { enumerable: true, get: function () { return PropertyInspector_js_1.LogPropertyInspector; } });
var MethodInspector_js_1 = require("./MethodInspector.js");
Object.defineProperty(exports, "MethodInspector", { enumerable: true, get: function () { return MethodInspector_js_1.MethodInspector; } });
Object.defineProperty(exports, "LogMethodInspector", { enumerable: true, get: function () { return MethodInspector_js_1.LogMethodInspector; } });
var ParameterInspector_js_1 = require("./ParameterInspector.js");
Object.defineProperty(exports, "ParameterInspector", { enumerable: true, get: function () { return ParameterInspector_js_1.ParameterInspector; } });
Object.defineProperty(exports, "LogParameterInspector", { enumerable: true, get: function () { return ParameterInspector_js_1.LogParameterInspector; } });
const isset = (val) => {
    return typeof val !== 'undefined' && val !== null;
};
const notset = (val) => {
    return (typeof val === 'undefined') || (val === null);
};
const isClassDecorator = (target, propertyKey, descriptor) => {
    return (isset(target)
        && notset(propertyKey)
        && notset(descriptor));
};
exports.isClassDecorator = isClassDecorator;
const isPropertyDecorator = (target, propertyKey, descriptor) => {
    return (isset(target)
        && isset(propertyKey)
        && notset(descriptor));
};
exports.isPropertyDecorator = isPropertyDecorator;
const isParameterDecorator = (target, propertyKey, descriptor) => {
    return (isset(target)
        && isset(propertyKey)
        && isset(descriptor)
        && typeof descriptor === 'number');
};
exports.isParameterDecorator = isParameterDecorator;
const isMethodDecorator = (target, propertyKey, descriptor) => {
    if ((isset(target)
        && isset(propertyKey)
        && isset(descriptor)
        && typeof descriptor === 'object')) {
        const propdesc = descriptor;
        return (typeof propdesc.value === 'function');
    }
    else {
        return false;
    }
};
exports.isMethodDecorator = isMethodDecorator;
const isAccessorDecorator = (target, propertyKey, descriptor) => {
    if ((isset(target)
        && isset(propertyKey)
        && isset(descriptor)
        && typeof descriptor === 'object')) {
        const propdesc = descriptor;
        return (typeof propdesc.value !== 'function')
            && (typeof propdesc.get === 'function'
                || typeof propdesc.set === 'function');
    }
    else {
        return false;
    }
};
exports.isAccessorDecorator = isAccessorDecorator;
